### DISCUSSION TITLE
Connecting to database with Psycopg2 and FastAPI Routers

### DISCUSSION THREAD

**Discussion by BenL-github** [fastapi/fastapi#6222] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#
# database.py
#

class Database:
    def __init__(self):
        self.conn = self.connect()

    def connect(self):
        """
        Connect to database and return connection
        """
        print("Connecting to PostgreSQL Database...")
        try:
            load_dotenv()
            conn = psycopg2.connect(
                    host = os.getenv("POSTGRES_HOST"),
                    dbname = os.getenv("POSTGRES_DB"),
                    user = os.getenv("POSTGRES_USER"),
                    password = os.getenv("POSTGRES_PASSWORD"),
                    port = os.getenv("POSTGRES_PORT")
                )
        except psycopg2.OperationalError as e:
            print(f"Could not connect to Database: {e}")
            sys.exit(1)

        return conn

    def get_table(self):
        ...

#
# main.py
#

from fastapi import FastAPI
from database import Database
from routes import test_route

db = Database()
app = FastAPI(openapi_tags = tags_metadata)
app.include_router(test_route.router)

#
# routes/test_route.py
#

from fastapi import APIRouter
import main

router = APIRouter()

@router.get("/test_route")
def get_test_route():
    test_table = main.db.get_table()
    return users
```

### Description

Hi, the docs only go into detail with SQLAlchemy and not other methods of handling a database. I was wondering how I should handle psycopg2 with routers. I have an object called Database in main.py that is connected to the database with psycopg2. Currently I have each route importing main to access the database object. Is this the best way to do it? The way I have it now seems to create a circular dependency (main imports test_route, and test_route imports main) but it seems as though uvicorn does not care. 

### Operating System

macOS

### FastAPI Version

0.78.0

### Python Version

3.8.9

**Comment by JarroVGIT** [comment, marked as answer]
I don't want to presume to say what the _best_ way is, but a better way might be to put `db = Database()` in database.py, make it a global variable, and import it from there. Then you are using the same instance in all your imports. See [this](https://stackoverflow.com/a/13034908/1557060) answer on SO. 

That way, you are indeed preventing circulair imports. Hope this helps!

**Comment by BenL-github** [comment]
> I don't want to presume to say what the _best_ way is, but a better way might be to put `db = Database()` in database.py, make it a global variable, and import it from there. Then you are using the same instance in all your imports. See [this](https://stackoverflow.com/a/13034908/1557060) answer on SO.
> 
> That way, you are indeed preventing circulair imports. Hope this helps!

Thank you, this is indeed a better way! If anybody has other suggestions, let me know.

**Comment by rafaelmarino** [comment]
Hey @BenL-github, sorry this doesn't answer your question&mdash;and I'm asking one of my own instead: is your API to interact with external users? If so, how do you plan on handling sql injection? Even though I'm more comfortable with psycopg2 for Postgres CRUD than an ORM, I feel like SQLAlchemy offers more of a "plug-and-play" approach to defend vs sql injection (vs mannually sanitizing queries in psycopg2). I'm having this dilemma right now...

**Comment by BenL-github** [comment]
@rafaelmarino 

psycopg2 goes into this: [link](https://www.psycopg.org/docs/usage.html#passing-parameters-to-sql-queries)

I'm not an expert on preventing injections but I think this may be enough
